design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+
+**Authors**: Aleksandr Sorokoumov (@gerrrr), Colin Hicks (@colinhicks)
+**Release Target**: TBD
+**Status**: _In Discussion_ |
+**Discussion**: [GitHub PR](https://github.com/confluentinc/ksql/pull/8550)
+
+## Motivation and background
+
+Many ksqlDB users leverage JSON. ETL pipelines often deal in large JSON blobs from, e.g., MongoDB.
+As the de-facto format of the web, it is often the format of analytics events produced to Kafka, as
+well as a choice for materialized views consumed by web apps.
+
+We want to position ksqlDB as a JSON-friendly database. Support for reading, writing, and
+traversing data as JSON presents an important value-add. And users of traditional databases like
+PostgreSQL already enjoy powerful support for JSON.
+
+Today, ksqlDB can:
+* Read JSON-formatted keys and values in input records.
+* Serialize keys and values as JSON in output records.
+* Extract a string value at a given JSON path.
+
+
+## What is in scope
+
+* Validate if a given string contains JSON value
+* Calculate the length of a JSON array encoded in a string
+* Extract keys and records from a given JSON-formatted string
+* Concatenate JSON-formatted strings
+* Convert a ksqlDB record to JSON-formatted string
+
+## What is not in scope
+
+* A JSON datatype
+* Binary JSON support
+
+## Public APIs
+
+Users can use ksqlDB’s varchar type on columns that they know are JSON formatted. A JSON-formatted
+string can also be nested within a map or struct.
+
+### is_json
+
+```
+is_json(json_string)
+```
+
+Given a string, returns true if it can be parsed as a valid JSON value, false otherwise.
+
+#### Examples
+
+```
+is_json("[1, 2, 3]") // returns true
+is_json("{}") // returns true
+is_json("1") // returns true
+is_json("\"abc\"") // returns true
+is_json("null") // returns true
+is_json("") // returns false
+is_json("abc") // returns false
+is_json(NULL) // returns false
+```
+
+
+### json_array_length
+
+```
+json_array_length(json_string)
+```
+
+Given a string, parses it as a JSON value and returns the length of the top-level array. Returns
+`NULL` if the string can't be interpreted as a JSON array, i.e., it is `NULL` or it does not contain
+valid JSON, or the JSON value is not an array.
+
+#### Examples
+
+```
+json_array_length("[1, 2, 3]") // returns 3
+json_array_length("[1, [1, [2]], 3]") // returns 3
+json_array_length("[]") // returns 0
+json_array_length("{}") // returns NULL
+json_array_length("123") // returns NULL
+json_array_length("abc") // returns NULL
+json_array_length(NULL) // returns NULL
+```
+
+### json_keys
+
+```
+json_keys(json_string)
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of strings representing the
+top-level keys. Returns `NULL` if the string can't be interpreted as a JSON object, i.e., it is
+`NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_keys("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns ["a", "b", "c"]
```

Comment:
Should this return `["a", "b", "d"]` ? `c` is the nested key?

Reply:
Good catch! Fixed in https://github.com/confluentinc/ksql/pull/8550/commits/5185e67e123fd03445abba27e8ade743b57eb07c.